### PR TITLE
S3 Support

### PR DESCRIPTION
**Pull request by lesserwhirls**
Add support for reading S3 objects as RandomAccessFiles, based off of the work done at https://github.com/Unidata/thredds/tree/feature/s3+hdfs

---
This change is [Reviewable](https://reviewable.io/reviews/unidata/netcdf-java/173)

### CODE REVIEW

**Comment by lesserwhirls** [comment]
Addresses the S3 part of Unidata/netcdf-java#111. Dependency size is a concern here due to pulling in the AWS S3 SDK (although thanks to v2.x of the SDK, we can just grab the S3 parts, so better than it was). Something to think about would be to make a RandomAccessFile Service Provider and split this out. I am guessing HDFS support will also be beefy. Maybe the same for Google Cloud Storage as well.

**Comment by lesserwhirls** [comment]
Looks like another timeout issue on the cache tests:

~~~bash
ucar.nc2.util.cache.TestFileCacheConcurrent > testConcurrentAccess STANDARD_OUT
    TestFileCacheConcurrent
     loaded 65 files
     submit 100 queue size 50 cache: {  hits= 0 miss= 33 nfiles= 31 elems= 16
    }
     done 100
     submit 200 queue size 53 cache: {  hits= 20 miss= 127 nfiles= 127 elems= 40
    }
     done 200
     submit 300 queue size 54 cache: {  hits= 80 miss= 166 nfiles= 50 elems= 19
    }
     done 300
     submit 400 queue size 44 cache: {  hits= 113 miss= 243 nfiles= 126 elems= 38
    }
     done 400
     submit 500 queue size 47 cache: {  hits= 160 miss= 293 nfiles= 50 elems= 21
    }
     done 500
     submit 600 queue size 48 cache: {  hits= 190 miss= 362 nfiles= 119 elems= 41
    }
     done 600
     submit 700 queue size 49 cache: {  hits= 233 miss= 418 nfiles= 50 elems= 19
    }
     done 700
     submit 800 queue size 49 cache: {  hits= 250 miss= 500 nfiles= 132 elems= 39
    }
     done 800
     submit 900 queue size 49 cache: {  hits= 290 miss= 561 nfiles= 70 elems= 24
    }
     done 900
     submit 1000 queue size 15 cache: {  hits= 334 miss= 651 nfiles= 158 elems= 47
    }
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted

ucar.nc2.util.cache.TestNetcdfFileCache STANDARD_OUT
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted
     InterruptedException=sleep interrupted

ucar.nc2.util.cache.TestNetcdfFileCache > testPeriodicClear STANDARD_OUT
     InterruptedException=sleep interrupted
~~~

**Comment by lesserwhirls** [comment]
> Something to think about would be to make a RandomAccessFile Service Provider and split this out. I am guessing HDFS support will also be beefy. Maybe the same for Google Cloud Storage as well.

So the `RandomAccessFile` Service Provider (`ucar.unidata.io.spi.RandomAccessFileProvider`) is part of this PR now, and S3 support has been broken out into its own module (`cdm-s3`). I imagine other S3 specific things will live here as well. Not sure if it would be appropriate for netCDF-Java to hold the code that generates a client catalog from an S3 listing, for example, but maybe.

File: cdm/core/src/main/java/ucar/nc2/NetcdfFiles.java [hunk 1]
```diff
@@ -290,31 +364,40 @@ else if (uriString.startsWith("file:"))
...
-      raf = new InMemoryRandomAccessFile(uriString, contents);
-    } else if (uriString.startsWith("s3:")) {
-      raf = new S3RandomAccessFile(uriString);
-      if (isCompressed(uriString))
-        raf = downloadAndDecompress(raf, uriString, buffer_size);
-    } else {
+    ucar.unidata.io.RandomAccessFile raf = null;
+
+    for (RandomAccessFileProvider provider : registeredRandomAccessFileProviders) {
+      if (provider.isOwnerOf("location")) {
```
Comment (JohnLCaron) [inline cdm/core/src/main/java/ucar/nc2/NetcdfFiles.java]: is this a typo?

if (provider.isOwnerOf("location"))

**Comment by JohnLCaron** [comment]
Would be good to cache at least the first block(s) for remote random access files, since the iosps all need to read from it. Might be well worth having a LoadingCache that is part of the RAF.

File: cdm/core/src/main/java/ucar/nc2/NetcdfFiles.java (hunk 1 above)
Comment (lesserwhirls) [reply cdm/core/src/main/java/ucar/nc2/NetcdfFiles.java]: Yep, total typo. Good catch.

**Comment by lesserwhirls** [comment]
> Would be good to cache at least the first block(s) for remote random access files, since the iosps all need to read from it. Might be well worth having a LoadingCache that is part of the RAF.

So there is a FIFO cache for blocks. As we try IOSPs, it seems like the first few blocks would get added the cache after the first IOSP is checked. Those would stay in the cache until it is filled, and then would be evicted. Are you saying we should modify that such that the first few blocks are always in the cache and do not get evicted?

**Review by JohnLCaron — COMMENTED** [review]
> FIFO cache for blocks.
Where is the FIFO cache for blocks?

> *[Reviewable](https://reviewable.io/reviews/unidata/netcdf-java/173#-:-Lzwgpx_1TLm0AkFrxzH:b-ibgte6)* status: 0 of 21 files reviewed, 2 unresolved discussions (waiting on @lesserwhirls)

*[a discussion](https://reviewable.io/reviews/unidata/netcdf-java/173#-LzwgbVhDHi52GGaXbyk:-LzwgbVhDHi52GGaXbyl:b-fixoi0) (no related file):*
This looks like a nice addition. Do we have a coverage test to make sure the new code in the non-S2 RAFs are being tested?

---
___
*[cdm/core/src/main/java/ucar/nc2/NetcdfFiles.java, line 195 at r1](https://reviewable.io/reviews/unidata/netcdf-java/173#-Lzwf6rcCIFUeA1-6rRp:-Lzwf6rcCIFUeA1-6rRq:bn7kp1) ([raw file](https://github.com/unidata/netcdf-java/blob/6904c5d7aa3f8106e4316af8467233dd4be972a4/cdm/core/src/main/java/ucar/nc2/NetcdfFiles.java#L195)):*
> ```
> private
> ```
Follow proscribed modifier order (see https://rules.sonarsource.com/java/tag/convention/RSPEC-1124) in new code.
I know its inconsistent in existing code, so try to clean up in the code that you touch when you have time.

**Review by lesserwhirls — COMMENTED** [review]
Right now it's down in cdm/s3/src/main/java/ucar/unidata/io/s3/S3RandomAccessFile.java, and pretty intimately tied to that class. I've refactored that out in my local branch, and made it to where HTTPRandomAccessFile takes advantage of it. Now I'm looking into making it use a Guava LoadingCache. I don't know that Raf is the right level to do big caches of reads (the existing Cache in Variable is probably right), but having a cache for the first MB or so for the IOSP checking process is worth it. Right now, the cache just hangs around until the Raf is closed, but with Guava we can give it a time to live, which I think is a nice feature for this purpose.

> *[Reviewable](https://reviewable.io/reviews/unidata/netcdf-java/173#-:-M-FY9Nu0F8tccR96L98:b85cbj7)* status: 0 of 21 files reviewed, 2 unresolved discussions (waiting on @JohnLCaron)

*[a discussion](https://reviewable.io/reviews/unidata/netcdf-java/173#-LzwgbVhDHi52GGaXbyk:-M-FZTPn6Jplef1oRtIo:b-r17ftl) (no related file):*
<details><summary><i>Previously, JohnLCaron (John Caron) wrote…</i></summary><blockquote>

This looks like a nice addition. Do we have a coverage test to make sure the new code in the non-S2 RAFs are being tested?
</blockquote></details>

I need to run this on Jenkins and look at the coverage...if not, I'll make sure the code is tested before merging.

---
___
*[cdm/core/src/main/java/ucar/nc2/NetcdfFiles.java, line 195 at r1](https://reviewable.io/reviews/unidata/netcdf-java/173#-Lzwf6rcCIFUeA1-6rRp:-M-FZbwh-tYE7xkdX4VQ:bq8qig3) ([raw file](https://github.com/unidata/netcdf-java/blob/6904c5d7aa3f8106e4316af8467233dd4be972a4/cdm/core/src/main/java/ucar/nc2/NetcdfFiles.java#L195)):*
<details><summary><i>Previously, JohnLCaron (John Caron) wrote…</i></summary><blockquote>

> ```
> private
> ```
Follow proscribed modifier order (see https://rules.sonarsource.com/java/tag/convention/RSPEC-1124) in new code.
I know its inconsistent in existing code, so try to clean up in the code that you touch when you have time. 

</blockquote></details>

Oops...I'll make sure those are cleaned up as well. Thanks!

**Comment by lesserwhirls** [comment]
Thanks for the feedback @JohnLCaron! I'll get those addressed soon.

**Comment by lesserwhirls** [comment]
Now both `S3RandomAccessFile` and `HTTPRandomAccessFile` have a Guava `LoadingCache` thanks to a new `RemoteRandomAccessFile` abstract class. Other remote random access file implementations will be able to take advantage of that, too.  Will run this on Jenkins to see what the coverage looks like and see what breaks.

**Comment by lesserwhirls** [comment]
I have a few fixes and a few more tests to get in here, but then we should be good.

**Comment by lesserwhirls** [comment]
The Jenkins tests passed, but only because we are missing some tests.

**Comment by lesserwhirls** [comment]
Added a fix for `InMemoryRandomAccessFile` and a test for `slurp://` (was not tested before). Also, now have a test that compares `NetcdfFile` objects produced from `RandomAccessFile`, `HTTPRandomAccessFile`, `InMemoryRandomAccessFile`, and `S3RandomAccessFile` with the same source file. I believe we're good now. Will give one more spin on Jenkins.

**Review by JohnLCaron — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/unidata/netcdf-java/173)* status: 0 of 24 files reviewed, 3 unresolved discussions (waiting on @JohnLCaron and @lesserwhirls)

___
*[cdm/core/build.gradle, line 27 at r3](https://reviewable.io/reviews/unidata/netcdf-java/173#-M-votqo0E0309WwicEx:-M-votqo0E0309WwicEy:b-t45xqp) ([raw file](https://github.com/unidata/netcdf-java/blob/973b36e47d8156f007c4718ad631a2d5c2a513f8/cdm/core/build.gradle#L27)):*
> ```
>   compile libraries["aws-s3"]
> ```

why does core depend on aws-s3? im hoping thats wrong.

**Review by lesserwhirls — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/unidata/netcdf-java/173)* status: 0 of 24 files reviewed, 3 unresolved discussions (waiting on @JohnLCaron)

___
*[cdm/core/build.gradle, line 27 at r3](https://reviewable.io/reviews/unidata/netcdf-java/173#-M-votqo0E0309WwicEx:-M-vv7qW3alyr9R9EH53:bkyvjnn) ([raw file](https://github.com/unidata/netcdf-java/blob/973b36e47d8156f007c4718ad631a2d5c2a513f8/cdm/core/build.gradle#L27)):*
<details><summary><i>Previously, JohnLCaron (John Caron) wrote…</i></summary><blockquote>

> ```
>   compile libraries["aws-s3"]
> ```

why does core depend on aws-s3? im hoping thats wrong.
</blockquote></details>

Yes, that's wrong - yikes! Good catch. Should be in cdm-s3. Fixed.

**Review by lesserwhirls — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/unidata/netcdf-java/173)* status: 0 of 24 files reviewed, 3 unresolved discussions (waiting on @JohnLCaron)

___
*[cdm/core/build.gradle, line 27 at r3](https://reviewable.io/reviews/unidata/netcdf-java/173#-M-votqo0E0309WwicEx:-M-vwzxND--bs5Vuv8mH:b-896fix) ([raw file](https://github.com/unidata/netcdf-java/blob/973b36e47d8156f007c4718ad631a2d5c2a513f8/cdm/core/build.gradle#L27)):*
<details><summary><i>Previously, lesserwhirls (Sean Arms) wrote…</i></summary><blockquote>

Yes, that's wrong - yikes! Good catch. Should be in cdm-s3. Fixed.
</blockquote></details>

Done.
___
*[cdm/core/src/main/java/ucar/nc2/NetcdfFiles.java, line 195 at r1](https://reviewable.io/reviews/unidata/netcdf-java/173#-Lzwf6rcCIFUeA1-6rRp:-M-vx33H6Qe-ac7wx7ff:b-896fix) ([raw file](https://github.com/unidata/netcdf-java/blob/6904c5d7aa3f8106e4316af8467233dd4be972a4/cdm/core/src/main/java/ucar/nc2/NetcdfFiles.java#L195)):*
<details><summary><i>Previously, lesserwhirls (Sean Arms) wrote…</i></summary><blockquote>

Oops...I'll make sure those are cleaned up as well. Thanks!
</blockquote></details>

Done.

**Comment by lesserwhirls** [comment]
Ok, I think we're good to go on this one, at least in terms of code testing, tests passing, etc.

**Comment by lesserwhirls** [comment]
Going to go ahead and merge this one in. w00t.

**Comment by pedro-cf** [comment]
Hello could I possibly create a Thredds catalog with a datasetScan that connects to a s3 bucket ?

**Comment by tdrwenski** [comment]
Hi @pedro-cf , here is an [example](https://github.com/Unidata/tds/blob/main/tds/src/test/content/thredds/tds-s3.xml) from our tests:
``` 
<datasetScan name="Test S3 dataset scan"
    ID="testS3DatasetScan"
    path="s3-dataset-scan"
    location="cdms3:thredds-test-data#delimiter=/">
    <serviceName>all</serviceName>
</datasetScan>
```

The location should be the cdms3 url of the form `cdms3:myBucket#delimiter=/` or `cdms3:myBucket?myKey/Prefix/#delimiter=/` (more info [here](https://docs.unidata.ucar.edu/netcdf-java/5.5/userguide/dataset_urls.html#object-stores)). Note the fragment `#delimiter=/` which is necessary to treat the object store buckets/key as hierarchical, like a file system.

If you have further questions feel free to open a new issue to discuss it!

**Comment by pedro-cf** [comment]
> Hi @pedro-cf , here is an [example](https://github.com/Unidata/tds/blob/main/tds/src/test/content/thredds/tds-s3.xml) from our tests:
> 
> ```
> <datasetScan name="Test S3 dataset scan"
>     ID="testS3DatasetScan"
>     path="s3-dataset-scan"
>     location="cdms3:thredds-test-data#delimiter=/">
>     <serviceName>all</serviceName>
> </datasetScan>
> ```
> 
> The location should be the cdms3 url of the form `cdms3:myBucket#delimiter=/` or `cdms3:myBucket?myKey/Prefix/#delimiter=/` (more info [here](https://docs.unidata.ucar.edu/netcdf-java/5.5/userguide/dataset_urls.html#object-stores)). Note the fragment `#delimiter=/` which is necessary to treat the object store buckets/key as hierarchical, like a file system.
> 
> If you have further questions feel free to open a new issue to discuss it!

https://github.com/Unidata/netcdf-java/issues/1191

**Pull request merged**